common/funding_tx.c

```diff
@@ -53,3 +56,271 @@ struct bitcoin_tx *funding_tx(const tal_t *ctx,
 	assert(bitcoin_tx_check(tx));
 	return tx;
 }
+
+#if EXPERIMENTAL_FEATURES
+/* We leave out the change addresses if there's no change left after fees */
+static size_t calculate_input_weights(struct input_info **inputs,
+				      struct amount_sat *total)
+{
+	u32 input_weight;
+	u64 scriptlen;
+	size_t weight = 0, i = 0;
+
+	*total = AMOUNT_SAT(0);
+	for (i = 0; i < tal_count(inputs); i++) {
+		/* prev_out hash + index + sequence */
+		input_weight = (32 + 4 + 4) * 4;
+
+		if (inputs[i]->script) {
+			scriptlen = tal_bytelen(inputs[i]->script);
+			input_weight += (scriptlen + varint_size(scriptlen)) * 4;
+		} else {
+			/* 00 byte script_sig len */
+			input_weight += 1 * 4;
+		}
+
+		input_weight += inputs[i]->max_witness_len;
+		weight += input_weight;
+
+		assert(amount_sat_add(total, *total, inputs[i]->input_satoshis));
+	}
+
+	return weight;
+}
+
+static size_t calculate_output_weights(struct output_info **outputs)
+{
+	size_t i, output_weights = 0, scriptlen;
+
+	for (i = 0; i < tal_count(outputs); i++) {
+		scriptlen = tal_bytelen(outputs[i]->script);
+		/* amount field + script + scriptlen varint */
+		output_weights += (8 + scriptlen + varint_size(scriptlen)) * 4;
+	}
+
+	return output_weights;
+}
+
+static size_t calculate_weight(struct input_info **opener_inputs,
+		               struct input_info **accepter_inputs,
+		               struct output_info **opener_outputs,
+		               struct output_info **accepter_outputs,
+			       struct amount_sat *opener_total,
+			       struct amount_sat *accepter_total)
+
+{
+	size_t weight;
+
+	/* version, input count, output count, locktime */
+	weight = (4 + 1 + 1 + 4) * 4;
+
+	/* add segwit fields: marker + flag */
+	weight += 1 + 1;
+
+	weight += calculate_input_weights(opener_inputs, opener_total);
+	weight += calculate_input_weights(accepter_inputs, accepter_total);
+
+	/* channel funding output: amount, len, scriptpubkey */
+	weight += (8 + 1 + BITCOIN_SCRIPTPUBKEY_P2WSH_LEN) * 4;
+
+	weight += calculate_output_weights(opener_outputs);
+	weight += calculate_output_weights(accepter_outputs);
+
+	return weight;
+}
+
+static const struct output_info *find_change_output(struct output_info **outputs)
+{
+	size_t i = 0;
+	for (i = 0; i < tal_count(outputs); i++) {
+		if (amount_sat_eq(outputs[i]->output_satoshis, AMOUNT_SAT(0)))
+			return outputs[i];
+	}
```

Comment:
And if they were to give us a 0-value output? That's certainly possible...

Reply:
are `OP_RETURN` outputs zero-value? :thinking: